rasa_core/policies/embedding_policy.py

```diff
@@ -0,0 +1,2109 @@
+                return tf.contrib.framework.with_same_shape(old, new)
+            return new
+
+        return tf.contrib.framework.nest.map_structure(
+                with_same_shape,
+                self,
+                super(TimeAttentionWrapperState, self)._replace(**kwargs)
+        )
+
+
+class TimeAttentionWrapper(tf.contrib.seq2seq.AttentionWrapper):
+    """Custom AttentionWrapper that takes into account time
+        when calculating attention.
+        Attention is calculated before calling rnn cell.
+
+        Modified from tensorflow's tf.contrib.seq2seq.AttentionWrapper.
+
+        Additional args:
+            attn_shift_range: Python int (`0` by default),
+                a time range within which to attend to the memory by location
+                by Neural Turing Machine.
+            sparse_attention: Python bool,
+                flag to use sparsemax (if `True`) instead of
+                softmax (if `False`, default) for probabilities
+            attn_to_copy_fn: (optional) a `callable`.
+                a function that picks which part of attention tensor
+                to use for copying to output, the default is `None`, which
+                turns off copying mechanism.
+                Copy inspired by: https://arxiv.org/pdf/1603.06393.pdf
+            output_attention: Python bool.  If `True`, the output at each
+                time step is the concatenated cell outputs, attention values
+                and ..., used in copy mechanism.
+
+        See the super class for other arguments description"""
+
+    def __init__(self, cell,
+                 attention_mechanism,
+                 dialogue_len,
+                 attn_shift_range=None,
+                 sparse_attention=False,
+                 attention_layer_size=None,
+                 alignment_history=False,
+                 attn_input_fn=None,
+                 cell_input_fn=None,
+                 index_of_attn_to_copy=None,
+                 similarity_fn=None,
+                 ignore_mask=None,
+                 emb_for_action_listen=None,
+                 output_attention=False,
+                 initial_cell_state=None,
+                 name=None):
+        super(TimeAttentionWrapper, self).__init__(
+                cell,
+                attention_mechanism,
+                attention_layer_size,
+                alignment_history,
+                cell_input_fn,
+                output_attention,
+                initial_cell_state,
+                name
+        )
+        self._attn_input_fn = attn_input_fn
+
+        self._dialogue_len = dialogue_len
+        if not isinstance(attn_shift_range, list):
+            attn_shift_range = [attn_shift_range]
+        self._timed_ntms = [TimedNTM(attn_shift_range[0],
+                                     sparse_attention,
+                                     name='0')]
+        if self._is_multi:
+            for i in range(1, len(attenti
```

Comment:
types? Or are we skipping this since it's taken from tf?

Reply:
yeah, I'd skip types here, they don't have it in `tf`. @tmbo ?

Reply:
yes thats fine :+1: